### PR TITLE
Calendar / Datepicker: a11y improvements, classes and icons options

### PR DESCRIPTION
**Pull request by fnagel**
This PR adds some missing a11y functionality to match the specs in wiki and updates the globalize / cldr dependencies.

Besides, I started implementing the classes option using `this._addClass` and `this._removeClass` methods. This required to change the HTML rendering from string concatenation to jQuery elements. For now I changed this for all elements except for the day rendering. For elements created once or a few times this should be no problem at all (header, header buttons, table header, etc.), but the day elements are rendered quite often so this could have some performance impact. 

Please give me some feedback on how to handle day elements and how to improve the classes implementation in general.

### CODE REVIEW

File: ui/widgets/calendar.js [hunk 1]
```diff
@@ -232,146 +239,157 @@ return $.widget( "ui.calendar", {
...
 			this.gridId = this.id + "-" + i;
+			labelledBy.push( this.gridId + "-title" );
 
-			labelledby.push( this.gridId + "-title" );
-			headerClass = "ui-calendar-header ui-widget-header ui-helper-clearfix";
+			html = $( "<div>" );
+			this._addClass( html, "ui-calendar-group" );
+
+			header = this._buildHeader();
+			this._removeClass( header, "ui-corner-all" );
```
Comment (fnagel) [inline ui/widgets/calendar.js]: I guess there is a better way to handle this using classes options?

**Comment by fnagel** [comment]
@scottgonzalez @arschmitz @jzaefferer Please take a look!

File: ui/widgets/calendar.js [hunk 2]
```diff
@@ -45,8 +45,18 @@ return $.widget( "ui.calendar", {
 	version: "@VERSION",
 	options: {
 		buttons: [],
+		classes: {
```
Comment (fnagel) [inline ui/widgets/calendar.js]: Please help what else classes we want to add here!

File: ui/widgets/calendar.js (hunk 1 above)
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: You'll need to use different classes for single and multi-picker instances. This will allow the default values for those to apply the correct corner classes. See `ui-selectmenu-button-open` and `ui-selectmenu-button-closed` for a similar example.

Comment (scottgonzalez) [inline ui/widgets/calendar.js]: Or more likely what we need in this case is three classes: first, middle, last. And we'll just apply first and last when it's single.

File: ui/widgets/calendar.js [hunk 3]
```diff
@@ -230,146 +243,158 @@ return $.widget( "ui.calendar", {
...
-				this._getTranslation( "datePickerRole" ) +
-			"</span>" +
-		"</div>";
+	_buildHeader: function() {
+		var header = $( "<div>" ),
+			title = $( "<div>", { role: "header", id: this.gridId + "-title" } ),
+			notice = $( "<span>" ).text( ", " + this._getTranslation( "datePickerRole" ) );
+
+		this._addClass( header, "ui-calendar-header", "ui-widget-header ui-helper-clearfix" )
+			._addClass( notice, "ui-helper-hidden-accessible" );
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: This is an extra class, not a keyed class.

File: ui/widgets/calendar.js [hunk 4]
```diff
@@ -230,146 +243,158 @@ return $.widget( "ui.calendar", {
...
+			month = $( "<span>" ).text( this.viewDate.monthName() ),
+			year = $( "<span>" ).text( this.viewDate.year() );
+
+		this._addClass( title, "ui-calendar-title" )
+			._addClass( month, "ui-calendar-month" )
+			._addClass( year, "ui-calendar-year" );
+
+		return title
+			.append( month )
+			.append( " " )
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: @timmywil Is this a safe use of `.append()`?

File: ui/widgets/calendar.js [hunk 5]
```diff
@@ -230,146 +243,158 @@ return $.widget( "ui.calendar", {
...
-			this._buildGridHeading() +
-			this._buildGridBody() +
-		"</table>";
+		var table = $( "<table>", {
+				role: "grid",
+				tabindex: 0,
+				"aria-readonly": true,
+				"aria-labelledby": this.gridId + "-month-label",
+				"aria-activedescendant": this._getDayId( this.date )
+			} );
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: Outdent one level.

File: ui/widgets/calendar.js (hunk 4 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: Info: needed to add some space between those elements

File: ui/widgets/calendar.js (hunk 5 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: Fixed.

File: ui/widgets/calendar.js (hunk 3 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: Changed.

File: ui/widgets/calendar.js (hunk 1 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: @scottgonzalez Added three classes as advised. Please take a look!

**Comment by fnagel** [comment]
@scottgonzalez @jzaefferer Updated branch, please take a look!

**Comment by fnagel** [comment]
I've got some more commits with added and improved unit tests and some minor bugfixes. I will add a new PR as soon this one has been merged.

@scottgonzalez @jzaefferer 
Question: in wiki there is a "Add change event (datepicker only)" todo item, but we never added this to the specs. Do we want this event? And do we really want to add it datepicker only when it needs to be implemented in calendar anyway (I'm pretty sure about this)?

Related: "Think about parameter for callback events" -- another one from the todo list which needs some more specification.

**Comment by scottgonzalez** [comment]
> Question: in wiki there is a "Add change event (datepicker only)" todo item, but we never added this to the specs. Do we want this event? And do we really want to add it datepicker only when it needs to be implemented in calendar anyway (I'm pretty sure about this)?

We have `select` which exists in both and is triggered only on a click or ENTER (actually, on `mousedown`, but I'm not sure why that is, seems like it should be on `click`)<sup>1</sup>. This will inform the user of all changes on the calendar. However, these events are not filtered, so selecting the currently selected date will result in another `select` event being triggered.

For datepicker, you can change the date without selecting a date from the calendar by typing a new date into the text input. On `blur`, we want to capture that the value has changed and trigger a `change` event. I assume this is why `change` is marked as datepicker only. I'm not strongly opposed to adding `change` to calendar, but it would be duplicative of `select` and would not have the additional useful cases that datepicker has.

On the other hand, we do have behavioral differences across widgets regarding programmatic value changes and triggering a `change` event. A widget like autocomplete which has no internal concept of a value will not trigger `change` on programmatic change of the input (because the value is changed outside of the widget). However, something like slider which has no concept of an external value will (because the value is changed inside the widget). So you could easily make the argument that calendar **should** have a `change` event and we should be triggering it whenever the value is changed programmatically. And if we're doing that, we should be triggering consistently when the value changes via any method.

SO IN CONCLUSION, let's add `change` to both :-)

_<sup>1</sup> We should also change ENTER to not trigger `mousedown`, but to instead just invoke a function which does the right thing._

**Comment by fnagel** [comment]
Thanks for your feedback! I will implement this soon.

> We should also change ENTER to not trigger mousedown, but to instead just invoke a function which does the right thing.

I actually already changed this in my latest local version due to inconsistent event return value and, well, because it looked just wrong :-D

File: ui/widgets/calendar.js (hunk 4 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: @timmywil Would you mind to give some feedback to the comment above? Thanks!

**Comment by fnagel** [comment]
@scottgonzalez @jzaefferer @arschmitz Any chance to get another review on this one?

And: any idea why there are no CI checks for this PR?

**Comment by scottgonzalez** [comment]
> And: any idea why there are no CI checks for this PR?

I think this is because @jzaefferer had accidentally turned Travis off for a while when he intended to turn it off for jqueryui.com. I bet it will work again on the next push.

File: ui/widgets/calendar.js (hunk 4 above)
Comment (timmywil) [inline ui/widgets/calendar.js]: I don't see a problem. This creates a text node and appends it.

File: ui/widgets/calendar.js (hunk 2 above)
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: This is fine if these are the only places we have style classes. We just need to document the full list of structural classes in the API docs.

File: ui/widgets/calendar.js [hunk 6]
```diff
@@ -230,146 +244,162 @@ return $.widget( "ui.calendar", {
...
+			labelledBy.push( this.gridId + "-title" );
 
-			labelledby.push( this.gridId + "-title" );
-			headerClass = "ui-calendar-header ui-widget-header ui-helper-clearfix";
-			if ( months[ i ].first ) {
-				headerClass += " ui-corner-left";
-			} else if ( months[ i ].last ) {
-				headerClass += " ui-corner-right";
-			}
+			html = $( "<div>" );
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: This variable name doesn't really make sense anymore.

File: ui/widgets/calendar.js [hunk 7]
```diff
@@ -230,146 +244,162 @@ return $.widget( "ui.calendar", {
...
+			this._addClass( html, "ui-calendar-group" );
 
-			html += "<div class='ui-calendar-group'>";
-			html += "<div class='" + headerClass + "'>" +
-				this._buildTitlebar() + "</div>";
-			html += this._buildGrid() + "</div>";
+			header = this._buildHeader();
+			this._addClass( header, "ui-calendar-header-" + (
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: That last opening paren should probably be on the next line.

File: ui/widgets/calendar.js [hunk 8]
```diff
@@ -230,146 +244,162 @@ return $.widget( "ui.calendar", {
...
 
 	_buildGridHeaderCell: function( day ) {
-		return "<th role='columnheader' abbr='" + day.fullname + "' aria-label='" + day.fullname + "'>" +
-			"<span title='" + day.fullname + "'>" +
-				day.shortname +
-			"</span>" +
-		"</th>";
+		return $( "<th role='columnheader' abbr='" + day.fullname + "' aria-label='" + day.fullname + "'>" +
+			"<span title='" + day.fullname + "'>" + day.shortname + "</span>" +
+		"</th>" );
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: Seems like you converted all the other multi-attribute elements to use `$(elem, props)` instead of lots of concatenation for HTML.

File: ui/widgets/datepicker.js [hunk 9]
```diff
@@ -279,6 +275,10 @@ var widget = $.widget( "ui.datepicker", {
 
 		this.isOpen = true;
 		this._trigger( "open", event );
+
+		this._delay( function() {
+			this.calendarInstance.grid.focus();
+		}, 1 );
```
Comment (scottgonzalez) [inline ui/widgets/datepicker.js]: Drop the duration parameter, since this is the default behavior.

File: ui/widgets/calendar.js (hunk 8 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: I've converted all string concatenations where it was needed in order to use the classes option. I tried to convert as little as possible due to performance considerations. 

For example: day cell rendering is called quite a lot so this still uses string concatenation and has no support for `classes` option yet.

Do we want to use `$(elem, props)` for _all_ elements?

File: ui/widgets/datepicker.js (hunk 9 above)
Comment (fnagel) [inline ui/widgets/datepicker.js]: Fixed.

File: ui/widgets/calendar.js (hunk 7 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: Fixed.

File: ui/widgets/calendar.js (hunk 6 above)
Comment (fnagel) [inline ui/widgets/calendar.js]: Fixed.

File: ui/widgets/calendar.js (hunk 8 above)
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: > Do we want to use $(elem, props) for all elements?

No, we don't have a mandate on either approach. This just stuck out in the diff. It's fine to leave as is.

Comment (fnagel) [inline ui/widgets/calendar.js]: Ok, ready to merge then? And how to get rid of this "Changes requested" message?

File: ui/widgets/calendar.js [hunk 10]
```diff
@@ -230,146 +244,162 @@ return $.widget( "ui.calendar", {
...
-		this.prevButton = this.element.find( ".ui-calendar-prev" );
-		this.nextButton = this.element.find( ".ui-calendar-next" );
 		this._refreshHeaderButtons();
-
 		this._createButtonPane();
 
 		this.grid = this.element.find( ".ui-calendar-calendar" );
 	},
 
+	_builSinglePicker: function() {
```
Comment (scottgonzalez) [inline ui/widgets/calendar.js]: typo

Comment (fnagel) [inline ui/widgets/calendar.js]: Fixed.
